lib/web_ui/test/html/image_test.dart

```diff
@@ -0,0 +1,126 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:html';
+import 'dart:typed_data';
+
+import 'package:test/bootstrap/browser.dart';
+import 'package:test/test.dart';
+import 'package:ui/src/engine.dart';
+import 'package:ui/ui.dart' hide TextStyle;
+
+void main() {
+  internalBootstrapBrowserTest(() => testMain);
+}
+
+typedef _ListPredicate<T> = bool Function(List<T>);
+_ListPredicate<T> deepEqualList<T>(List<T> a) {
+  return (List<T> b) {
+    if (a.length != b.length)
+      return false;
+    for (int i = 0; i < a.length; i += 1) {
+      if (a[i] != b[i])
+        return false;
+    }
+    return true;
+  };
+}
+
+Matcher listEqual(List<int> source, {int tolerance = 0}) {
+  return predicate(
+    (List<int> target) {
+      if (source.length != target.length)
+        return false;
+      for (int i = 0; i < source.length; i += 1) {
+        if ((source[i] - target[i]).abs() > tolerance)
+          return false;
+      }
+      return true;
+    },
+    source.toString(),
+  );
+}
+
+// Converts `rawPixels` into a list of bytes that represent raw pixels in rgba8888.
+//
+// Each element of `rawPixels` represents a bytes in order 0xRRGGBBAA, with
+// pixel order Left to right, then top to bottom.
+Uint8List _pixelsToBytes(List<int> rawPixels) {
+  return Uint8List.fromList((() sync* {
+    for (final int pixel in rawPixels) {
+      yield (pixel >> 24) & 0xff; // r
+      yield (pixel >> 16) & 0xff; // g
+      yield (pixel >> 8)  & 0xff; // b
+      yield (pixel >> 0)  & 0xff; // a
+    }
+  })().toList());
+}
+
+Future<Image> _encodeToHtmlThenDecode(Uint8List rawBytes, int width, int height) async {
+  final ImageDescriptor descriptor = ImageDescriptor.raw(
+    await ImmutableBuffer.fromUint8List(rawBytes),
+    width: 2,
+    height: 2,
```

Comment:
Do the opacity tests get any better if you just do a 1x1 pixel image that's fully transparent?

Reply:
The test is fine if all pixels of the image is either fully opaque or fully transparent. The problem occurs if it contains any pixel between 1 and 254. So 1x1 fully transparent should work fine (and has been tested in the opaque image one.)